EIPS/eip-erc721-marketplace-extension.md

```diff
@@ -0,0 +1,323 @@
+
+## Motivation
+
+OpenSea’s latest code snippet gives them the ability to entirely control which
+platform your NFTs can (or cannot) be traded on. Our goal is to give that
+control back to the project creators.
+
+## Specification
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL
+NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY",
+and "OPTIONAL" in this document are to be interpreted as described in RFC 2119
+and RFC 8174.
+
+Implementers of this standard **MUST** have all of the following functions:
+
+```solidity
+pragma solidity ^0.8.0;
+import './IERC721.sol'
+
+///
+/// @dev Interface for the ERC721 Marketplace Extension
+///
+interface IERC721MarketplaceExtension {
+  ///
+  /// @dev A structure representing a listed token
+  ///
+  /// @param tokenId - the NFT asset being listed
+  /// @param tokenPrice - the price the token is being sold for, regardless of currency
+  /// @param to - address of who this listing is for, 
+  ///             can be address zero for a public listing,
+  ///             or non zero address for a private listing 
+  ///
+  struct Listing {
+    uint256 tokenId;
+    uint256 tokenPrice;
+    address to;
+  }
+  ///
+  /// @dev Emitted when a token is listed for sale.
+  ///
+  /// @param tokenId - the NFT asset being listed
+  /// @param from - address of who is selling the token
+  /// @param to - address of who this listing is for, 
+  ///             can be address zero for a public listing,
+  ///             or non zero address for a private listing 
+  /// @param price - the price the token is being sold for, regardless of currency
+  ///
+  event Listed( uint256 indexed tokenId, address indexed from, address to, uint256 indexed price );
+  ///
+  /// @dev Emitted when a token that was listed for sale is being delisted
+  ///
+  /// @param tokenId - the NFT asset being delisted
+  ///
+  event Delisted( uint256 indexed tokenId );
+  ///
+  /// @dev Emitted when a token that was listed for sale is being purchased.
+  ///
+  /// @param tokenId - the NFT asset being purchased
+  /// @param from - address of who is selling the token
+  /// @param to - address of who is buying the token 
+  /// @param price - the price the token is being sold for, regardless of currency
+  ///
+  event Purchased( uint256 indexed tokenId, address indexed from, address indexed to, uint256 price );
+
+  ///
+  /// @dev Lists token `tokenId` for sale.
+  ///
+  /// @param tokenId - the NFT asset being listed
+  /// @param to - address of who this listing is for, 
+  ///             can be address zero for a public listing,
+  ///             or non zero address for a private listing 
+  /// @param price - the price the token is being sold for, regardless of currency
+  ///
+  /// Requirements:
+  /// - `tokenId` must exist
+  /// - Caller must own `tokenId`
+  /// - Must emit a {Listed} event.
+  ///
+  function listItem( uint256 tokenId, uint256 price, address to ) external;
```

Comment:
I would like to see this feature support multiple currencies. 

Existing marketplaces support creating a sale in ETH or a stablecoin. It's common for NFT ecosystems to create their own tokens that they would like to use for this purpose. 

Adding a `paymentTokenAddress` would add arbitrary ERC20 support to the marketplace, allowing the seller to peg the price to any ERC20. 
This would also support projects that want to allow only a subset of ERC20s for sale as they can whitelist ERC20 addresses in their implementation. 
This doesn't preclude selling in ETH as `address(0)` can be used to indicate a sale price in ETH.

Reply:
Curious... I think it's a good idea in theory, but it would add some more to the codebase.

We intentionally left the feature-set a little light to leave expansion decisions on the Creator and they could add whatever they needed. Making it a requirement, if it could be done efficiently would be nice. I mean a simple check on the payment for eth sent with the transfer and if not then try to transferFrom with whatever contract information, sounds simple enough.

In that case, if the creator chose to implement those other tokens they could do so easily and if they decided not to then they could leave the base option out. ... 

I see some promise in this idea and have been thinking about it since I saw the post earlier, I was away from my computer and busy with other things. Definitely something worth considering. Thank you for the comment and we'll spend some time discussing this.

My main worry would be small artist collections trying to build out a front end to accomodate, but hopefully we can just adapt the broiler code as well easily enough.

Thanks again, will give this some more thought.

Reply:
That's a good point, might be worth implementing... 
Our initial thought was that the coin to pay with could be up to the project creator, but it would only allow for one type of payment to be used, whereas your idea allows for unlimited options, with more flexibility.

Reply:
Agree. EIP should be as concise and general as possible.